### DISCUSSION TITLE
custom V-USB library

### DISCUSSION THREAD

**Discussion by nerdralph** [SpenceKonde/ATTinyCore#526] [Development Plans]
I thought of a way of possibly getting V-USB to coexist with ATTinyCore.  Create a V-USB wrapper library that implments it's own millis using the SOFHOOK() in V-USB.  You may need to break up the code in wiring.c into separate files (make a millis.c with the millis() fn and the ISR) in order to make it easier to override in the V-USB library.

This would also cleanly solve the 16.5/16MHz millis problem when using the 16.5MHz V-USB, as the USB SOF keepalive pulses come in at a steady 1KHz.

**Comment by prandeamus** [comment]
I like this idea. I'm learning about how V-USB works even if I don't understand every detail. If you don't mind, here are some stream-of-consciousness notes from what passes for my brain

# Interupts and millis()

It seems that a common design pattern (from the code I've seen so far) in initialising is as follows

cli();
usbDeviceDisconnect();
_delay_ms(500);
usbDeviceConnect();
usbInit();
sei();

I believe this is also needed after a crash or other failure to make sure that both ends are in sync. It sometimes is associated with a watchdog reset. I guess if interrupts are disabled, you can't use any sort of regular delay() or millis() function, so it uses the compiler intrinsic for the half second delay. 

e.g. see https://github.com/heartscrytech/DigisparkMIDI/blob/master/DigiMIDI.h

Not sure if that affects your design choice about using SOF keepalive.

# SOF frequency 

Is SOF keepalive rate accurate to 1kHz, and if an app uses USB/SOF to implement mills()  you're making the accuracy dependent on the USB host. Not sure what the tolerances are.

# Calibration

I found this recently https://github.com/jokkebk/TrinketMIDI/blob/master/trinketusb.c and nd it has an optional OSCCAL calibration function, presumably for those devices without a crystal. I'm mentioning it on the off-chance that it is of use in this discussion. Even if the idea is sound, it uses runtime floating point operations so may not be desirable in this form.

Hope this stimulates debate, but please feel free to shout me down when I'm off-base, which does happen.

**Comment by nerdralph** [comment]
@prandeamus Here's a couple of posts that may help you with understanding USB:
http://nerdralph.blogspot.com/2015/01/usb-interfacing-for-avr-microcontrollers.html
http://nerdralph.blogspot.com/2020/08/hackers-intro-to-usb-hardware.html

Although my latest blog post is focused on the CH55x MCUs, much of it is relevant to USB in general:
http://nerdralph.blogspot.com/2021/03/writing-usb-firmware-on-ch55x-mcus.html

To specifically address your question about the SOF timing accuacy,  AFAIR the spec only requires +-0.25% for low-speed USB, but all PCs and laptops use xtal oscillators for the USB controllers, which have typical accuracy of +/-50ppm

If you want a *REALLY* solid understanding of USB, then read this site:
https://www.beyondlogic.org/usbnutshell/usb1.shtml
Read every chapter, and then read it as many times as it takes to make sense.

You can read about the latest and greatest OSCCAL tuner for V-USB and variants here:
https://github.com/micronucleus/micronucleus/discussions/216

**Reply by prandeamus** [reply]
Thanks :) I think.

**Comment by prandeamus** [comment]
Presumably, the core did use SOF as a timing source for millis() this would need to be dynamic/hotpluggable as we can't assume that a device will always be plugged into a USB host?
